lib/reservation_tracker.rb

```diff
@@ -0,0 +1,267 @@
+        reservation.room
+      end
+      return reserved_rooms
+    end
+
+    def find_blocks(requested_dates)
+      blocks_by_date = @blocks.map do |room|
+        date_range = room.date_range
+        room if date_range.overlaps?(requested_dates)
+      end
+      return blocks_by_date.compact
+    end
+
+    def find_unavailable_rooms(requested_dates)
+      reserved_rooms = find_reserved_rooms(requested_dates)
+      blocks = find_blocks(requested_dates)
+      unavailable_rooms = reserved_rooms + blocks
+      return unavailable_rooms
+    end
+
+    def find_available_rooms(requested_dates)
+      unavailable_rooms = find_unavailable_rooms(requested_dates)
+      available_rooms = @rooms.reject { |room| unavailable_rooms.include?(room) }
+      check_availablity?(available_rooms)
+      return available_rooms
+    end
+
+    def get_first_available_room(requested_dates)
+      all_available_rooms = find_available_rooms(requested_dates)
+      first_available_room = all_available_rooms.first
+      return first_available_room
+    end
+
+    def confirm_valid_dates?(start_date, end_date)
+      check_dates_validity?(start_date, end_date)
+      check_dates_order?(start_date, end_date)
+    end
+
+    def get_requested_dates(start_date, end_date)
+      check_prior_today?(start_date)
+      confirm_valid_dates?(start_date, end_date)
+      return DateRange.new(start_date, end_date)
+    end
+
+    def reserve_room(input)
+      requested_dates = get_requested_dates(input[:start_date], input[:end_date])
+
+      reservation_id = @reservations.length + 1
+      block_id = input[:block_id]
+      room = get_first_available_room(requested_dates)
+
+      reservation_data = {
+        id: reservation_id,
+        block_id: block_id,
+        room: room,
+        date_range: requested_dates
+      }
+
+      new_reservation = Reservation.new(reservation_data)
+      @reservations << new_reservation
+      return new_reservation
+    end
+
+    def confirm_valid_qty?(requested_qty)
+      check_valid_num?(requested_qty)
+      check_num_requested?(requested_qty)
+    end
+
+    def get_blocks(requested_qty, requested_dates)
+      confirm_valid_qty?(requested_qty)
+      available_rooms = find_available_rooms(requested_dates)
+      check_enough_rooms?(available_rooms, requested_qty)
+      block = available_rooms.take(requested_qty)
+      return block
+    end
+
+    def block_rooms(input)
+      requested_dates = get_requested_dates(input[:start_date], input[:end_date])
+      block_id = @blocks.length + 1
+      blocked_qty_rooms = get_blocks(input[:party], requested_dates)
+
+      block_data = {
+        id: block_id,
+        party: blocked_qty_rooms,
+        date_range: requested_dates
+      }
+
+      new_block = Block.new(block_data)
+      @blocks << new_block
+      return new_block
+    end
+
+    private
+
+    class OldStartDateError < StandardError; end
+    class InvalidDateError < StandardError; end
```

Comment:
I like that you made all these custom exceptions. You probably don't want to make them `private` though, because then people outside this class won't be able to access it (for example to rescue those specific exceptions).